tools/mcms_proposal/encoding.go

```diff
@@ -0,0 +1,159 @@
+package mcms_proposal
+
+import (
+	"math/big"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/errors"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/gethwrappers"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/merkle"
+)
+
+var MANY_CHAIN_MULTI_SIG_DOMAIN_SEPARATOR_OP = crypto.Keccak256Hash([]byte("MANY_CHAIN_MULTI_SIG_DOMAIN_SEPARATOR_OP"))
+var MANY_CHAIN_MULTI_SIG_DOMAIN_SEPARATOR_METADATA = crypto.Keccak256Hash([]byte("MANY_CHAIN_MULTI_SIG_DOMAIN_SEPARATOR_METADATA"))
+
+func calculateTransactionCounts(transactions []ChainOperation) map[string]uint64 {
+	txCounts := make(map[string]uint64)
+	for _, tx := range transactions {
+		txCounts[tx.ChainIdentifier]++
+	}
+	return txCounts
+}
+
+func buildRootMetadatas(
+	chainMetadata map[string]ChainMetadata,
+	txCounts map[string]uint64,
+	currentOpCounts map[string]big.Int,
+	overridePreviousRoot bool,
+) (map[string]gethwrappers.ManyChainMultiSigRootMetadata, error) {
+	rootMetadatas := make(map[string]gethwrappers.ManyChainMultiSigRootMetadata)
+
+	for chainID, metadata := range chainMetadata {
+		stringChainId, err := big.NewInt(0).SetString(chainID, 10)
+		if !err {
+			return nil, &errors.ErrInvalidChainID{
+				ReceivedChainID: chainID,
+			}
+		}
+
+		currentNonce, ok := currentOpCounts[chainID]
+		if !ok {
+			return nil, &errors.ErrMissingChainDetails{
+				ChainIdentifier: chainID,
+				Parameter:       "current op count",
+			}
+		}
+
+		currentTxCount, ok := txCounts[chainID]
+		if !ok {
+			return nil, &errors.ErrMissingChainDetails{
```

Comment:
I don't follow what you mean by this, can you elaborate?

Reply:
Maybe this is my lack of golang experience, I was thinking if there was a way to return the error object directly instead of the address of the error. 

However, the way you have it for custom errors seems idioimatic: https://gobyexample.com/custom-errors

Reply:
yeah I think this is just the way they are instantiated, because `error` is still the return time